common/app/layout/Front.scala

```diff
@@ -378,6 +377,121 @@ object Front extends implicits.Collections {
                       adFree: Boolean)(implicit requestHeader: RequestHeader): Front =
     Front(fromPressedPageWithDeduped(pressedPage, edition, initialContext, adFree))
 
+  def fromSectionPage(frontPage: SectionPage, edition: Edition)(implicit context: ApplicationContext): Front = {
+
+    def fastContainerWithMpu(numberOfItems: Int): Option[ContainerDefinition] = numberOfItems match {
+      case 2 => Some(FixedContainers.fastIndexPageMpuII)
+      case 4 => Some(FixedContainers.fastIndexPageMpuIV)
+      case 6 => Some(FixedContainers.fastIndexPageMpuVI)
+      case n if n >= 9 => Some(FixedContainers.fastIndexPageMpuIX)
+      case _ => None
+    }
+
+    def slowContainerWithMpu(numberOfItems: Int): Option[ContainerDefinition] = numberOfItems match {
+      case 2 => Some(FixedContainers.slowIndexPageMpuII)
+      case 4 => Some(FixedContainers.slowIndexPageMpuIV)
+      case 5 => Some(FixedContainers.slowIndexPageMpuV)
+      case 7 => Some(FixedContainers.slowIndexPageMpuVII)
+      case _ => None
+    }
+
+    @tailrec
+    def containerDefinition(groupings: Seq[FrontPageGrouping],
```

Comment:
This could probably be converted to a map since accumulation isn't used while recurring (just appended).
The the whole outer function is horrid. I can see it's been moved from IndexPage. Could refactor in the future!

Reply:
@philmcmahon and I thought about refactoring this huge and ugly function we moved from https://github.com/guardian/frontend/pull/18572/files#diff-ca2f0b1817d36aea771ec4339ee78fdf
However this PR is big enough without addding more refactoring in it. I would prefer if this could be done in a second time after this is merged.